# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [223]:
import pandas as pd
import sys
sys.path.append('../../Group10')
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [224]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [225]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [226]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [227]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [228]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [229]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [230]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [231]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [232]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [233]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [234]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [235]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [236]:
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [237]:
df['Wteam'].value_counts()

Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [238]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [239]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [240]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [241]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [242]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [243]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [244]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [245]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [246]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [247]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [248]:
df1 = df.sort_values('Season')
df2 = df.sort_values('Wteam')
df3 = df.sort_values('Wscore')

rows= df3.iloc[100:201, [0, 1, 2]]

print(rows)

        Season  Daynum  Wteam
99013     2008      33   1460
126840    2013      77   1366
126982    2013      80   1442
15619     1989      30   1343
122700    2012     103   1349
...        ...     ...    ...
135433    2015      26   1438
72338     2002      90   1443
135808    2015      36   1185
129196    2013     131   1380
132791    2014      92   1155

[101 rows x 3 columns]


**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [249]:
df1.iloc[[df1['Wscore'].argmax()]][['Season', 'Wteam']]


Season  Wteam
24970    1991   1258

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [250]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [251]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [252]:
df['ScoreDifference']=df['Wscore']-df['Lscore']
df[df['ScoreDifference'] > df['ScoreDifference'].mean()]


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [253]:
max_index = filtered_df['ScoreDifference'].idxmax()


season_position = filtered_df.columns.get_loc('Season')
Winning_pos = filtered_df.columns.get_loc('Wteam')
Loosing_pos = filtered_df.columns.get_loc('Lteam')

season = filtered_df.iloc[max_index, season_pos]
team1 = filtered_df.iloc[max_index, Winning_pos]
team2 = filtered_df.iloc[max_index, Loosing_pos]

print("Season:", season)
print("Teams:", team1, "vs", team2)

Season: 2010
Teams: 1330 vs 1241


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [254]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [255]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [256]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [257]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [258]:
df.groupby('Season')['Wscore'].mean()

print(average_winning_score)

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [259]:
df.groupby('Wteam')['Wscore'].max()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [260]:
df.groupby(['Season','Wteam'])['Wscore'].mean().max()

126.34782608695652

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [261]:
df.groupby(['Season', 'Wteam'])['Wteam'].count()

Season  Wteam
1985    1102      5
        1103      9
        1104     21
        1106     10
        1108     19
                 ..
2016    1460     20
        1461     12
        1462     27
        1463     21
        1464      9
Name: Wteam, Length: 10172, dtype: int64

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [262]:
# Write your code here


team_wins = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')


most_wins_per_season = team_wins.groupby('Season')['Wins'].idxmax()

teams_with_most_wins = team_wins.loc[most_wins_per_season]


print(teams_with_most_wins)

      Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [263]:
team_loses = df.groupby(['Season', 'Lteam']).size().reset_index(name='loses')


most_loses_per_season = team_loses.groupby('Season')['loses'].idxmax()

teams_with_most_loses = team_loses.loc[most_loses_per_season]


print(teams_with_most_loses)

#No, there's are teams that have a high loss score than other teams with wins.


      Season  Lteam  loses
82      1985   1209     25
408     1986   1264     26
771     1987   1361     25
948     1988   1218     25
1226    1989   1203     27
1609    1990   1317     25
2007    1991   1447     26
2050    1992   1138     26
2447    1993   1259     25
2674    1994   1170     26
2950    1995   1142     26
3261    1996   1152     25
3591    1997   1179     26
3850    1998   1126     26
4241    1999   1224     25
4531    2000   1194     28
4779    2001   1115     25
5097    2002   1115     26
5450    2003   1152     27
5875    2004   1259     27
6198    2005   1255     29
6513    2006   1239     26
6841    2007   1233     28
7254    2008   1312     29
7598    2009   1312     30
7754    2010   1108     29
8383    2011   1405     28
8729    2012   1406     31
8889    2013   1212     28
9402    2014   1381     27
9733    2015   1363     28
9877    2016   1152     28


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [264]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [279]:
high_scoring_game = []

for index, row in df.iterrows():
    # Check if the winning score is greater than 100
    if row['Wscore'] > 100:

        high_scoring_game.append('Yes')
    else:
        high_scoring_game.append('No')

df['HighScoringGame'] = high_scoring_game

print(df)







        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3              

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [283]:
import pandas as pd

# Create an empty dictionary to store the number of games for each team
team_games = {}

for _, row in df.iterrows():
    winning_team = row['Wteam']
    team_games[winning_team] = team_games.get(winning_team, 0) + 1
    
    # Get the losing team and update the count in the dictionary
    losing_team = row['Lteam']
    team_games[losing_team] = team_games.get(losing_team, 0) + 1

# Create a DataFrame from the dictionary
team_games_df = pd.DataFrame({'Team': list(team_games.keys()), 'Total Games': list(team_games.values())})


print(team_games_df)

     Team  Total Games
0    1228          992
1    1328          968
2    1106          855
3    1354          906
4    1112          981
..    ...          ...
359  1297          113
360  1213           84
361  1262           83
362  1101           76
363  1230           64

[364 rows x 2 columns]


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [284]:
import pandas as pd

# Create an empty dictionary to store the highest score difference for each season
highest_score_diff = {}

# Iterate over the rows of the DataFrame
for _, row in df.iterrows():
    # Get the season and calculate the score difference
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']
    
    if season in highest_score_diff:
        # Update the highest score difference if the current score difference is greater
        if score_diff > highest_score_diff[season]['ScoreDifference']:
            highest_score_diff[season] = {
                'Game': _,  # Store the row index of the game
                'ScoreDifference': score_diff
            }
    else:
        highest_score_diff[season] = {
            'Game': _,  # Store the row index of the game
            'ScoreDifference': score_diff
        }

# Create a DataFrame from the dictionary
highest_score_diff_df = pd.DataFrame.from_dict(highest_score_diff, orient='index')


print(highest_score_diff_df)

        Game  ScoreDifference
1985     236               60
1986    4731               84
1987    8240               73
1988   12046               68
1989   16677               70
1990   19502               76
1991   25161               68
1992   27997               82
1993   33858               81
1994   36404               69
1995   39858               74
1996   44653               91
1997   49033               80
1998   52600               66
1999   57618               75
2000   60616               75
2001   65999               75
2002   70737               59
2003   74791               72
2004   79181               70
2005   83372               64
2006   89542               72
2007   94507               74
2008   97737               70
2009  103723               68
2010  108405               75
2011  114015               69
2012  118699               62
2013  124057               58
2014  131373               74
2015  134590               69
2016  140040               73


Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [285]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')



**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [289]:


# Calculate the total number of games played by each team
win_counts = df['Wteam'].value_counts()
loss_counts = df['Lteam'].value_counts()

# Add the two Series together to get the total games count for each team
total_games_counts = win_counts.add(loss_counts, fill_value=0)

total_games_df = pd.DataFrame({'TotalGames': total_games_counts})

print(total_games_df)

      TotalGames
1101          76
1102         840
1103         910
1104         975
1105         447
...          ...
1460         827
1461         914
1462         954
1463         838
1464         856

[364 rows x 1 columns]


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [292]:

# Calculate the score difference using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

max_score_diff_indexes = df.groupby('Season')['ScoreDifference'].idxmax()

# Get the rows corresponding to the index of the game with the highest score difference for each season
max_score_diff_games = df.loc[max_score_diff_indexes]


print(max_score_diff_games)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410    

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [271]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [272]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [273]:
type(df['Wscore'])

pandas.core.series.Series

In [274]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [275]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [276]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [277]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y